code/modules/lighting/lighting_source.dm

```diff
@@ -0,0 +1,288 @@
+	var/atom/source_atom     // The atom that we belong to.
+
+	var/turf/source_turf     // The turf under the above.
+	var/light_power    // Intensity of the emitter light.
+	var/light_range      // The range of the emitted light.
+	var/light_color    // The colour of the light, string, decomposed by parse_light_color()
+
+	// Variables for keeping track of the colour.
+	var/lum_r
+	var/lum_g
+	var/lum_b
+
+	// The lumcount values used to apply the light.
+	var/tmp/applied_lum_r
+	var/tmp/applied_lum_g
+	var/tmp/applied_lum_b
+
+	var/list/datum/lighting_corner/effect_str     // List used to store how much we're affecting corners.
+	var/list/turf/affecting_turfs
+
+	var/applied = FALSE // Whether we have applied our light yet or not.
+
+	var/vis_update      // Whether we should smartly recalculate visibility. and then only update tiles that became (in)visible to us.
+	var/needs_update    // Whether we are queued for an update.
+	var/destroyed       // Whether we are destroyed and need to stop emitting light.
+	var/force_update
+
+/datum/light_source/New(var/atom/owner, var/atom/top)
+	source_atom = owner // Set our new owner.
+	if (!source_atom.light_sources)
+		source_atom.light_sources = list()
+
+	source_atom.light_sources += src // Add us to the lights of our owner.
+	top_atom = top
+	if (top_atom != source_atom)
+		if (!top.light_sources)
+			top.light_sources     = list()
+
+		top_atom.light_sources += src
+
+	source_turf = top_atom
+	light_power = source_atom.light_power
+	light_range = source_atom.light_range
+	light_color = source_atom.light_color
+
+	parse_light_color()
+
+	effect_str      = list()
+	affecting_turfs = list()
+
+	update()
+
+	return ..()
+
+// Kill ourselves.
+/datum/light_source/proc/destroy()
+	destroyed = TRUE
+	force_update()
+	if (source_atom)
+		source_atom.light_sources -= src
+
+	if (top_atom)
+		top_atom.light_sources    -= src
+
+// Fuck supporting force.
+/datum/light_source/Destroy(var/force)
+	destroy()
+	return QDEL_HINT_IWILLGC
+
+// Yes this doesn't align correctly on anything other than 4 width tabs.
+// If you want it to go switch everybody to elastic tab stops.
+// Actually that'd be great if you could!
+#define EFFECT_UPDATE                   \
+	if (!needs_update)                  \
+	{                                   \
+		lighting_update_lights += src;  \
+		needs_update            = TRUE; \
+	}
+
+// This proc will cause the light source to update the top atom, and add itself to the update queue.
+/datum/light_source/proc/update(var/atom/new_top_atom)
+	// This top atom is different.
+	if (new_top_atom && new_top_atom != top_atom)
+		if(top_atom != source_atom) // Remove ourselves from the light sources of that top atom.
+			top_atom.light_sources -= src
+
+		top_atom = new_top_atom
+
+		if (top_atom != source_atom)
+			if(!top_atom.light_sources)
+				top_atom.light_sources = list()
+
+			top_atom.light_sources += src // Add ourselves to the light sources of our new top atom.
```

Comment:
Replace this and the `if(!top_atom.light_sources)` bit with `LAZYADD(src, top_atom.light_sources)` which will do the same thing